### DISCUSSION TITLE
Simulation Study: SBERT models

### DISCUSSION THREAD

**Discussion by rohitgarud** [asreview/asreview#1405] [Show and tell]
The SBERT feature extraction strategy allows for use of different Sentence BERT models. The default is 'all-mpnet-base-v2'. There is another model by AllenAI which is specifically trained on scientific papers called [SPECTER](https://github.com/allenai/specter/tree/master). This simulation study was performed to check whether selecting the SPECTER model (['allenai-specter'](https://www.sbert.net/docs/pretrained_models.html)) as the transformer model for SBERT improves performance using the Similarity Classifier. The results are presented here

**Comment by rohitgarud** [comment]
| Dataset                 | Feature Extraction | Classifier  | recall_0.5 | wss_0.95 | erf_0.1 | atd      |
|-------------------------|--------------------|-------------|------------|----------|---------|----------|
| ACEInhibitors           | SBERT-mpnet        | Similarity  | 0.950      | 0.739    | 0.775   | 202.975  |
|                         | SBERT-specter      |             | 0.950      | 0.491    | 0.700   | 228.550  |
|                         | TFIDF              | Naïve Bayes | 0.950      | 0.782    | 0.750   | 190.325  |
| ADHD                    | SBERT-mpnet        | Similarity  | 0.947      | 0.747    | 0.632   | 94.842   |
|                         | SBERT-specter      |             | 0.947      | 0.554    | 0.737   | 84.263   |
|                         | TFIDF              | Naïve Bayes | 0.947      | 0.446    | 0.737   | 96.789   |
| Antihistamines          | SBERT-mpnet        | Similarity  | 0.867      | 0.214    | 0.400   | 77.200   |
|                         | SBERT-specter      |             | 0.867      | 0.341    | 0.267   | 80.533   |
|                         | TFIDF              | Naïve Bayes | 0.867      | 0.091    | 0.267   | 72.333   |
| Appenzeller-Herzog_2020 | SBERT-mpnet        | Similarity  | 0.964      | 0.563    | 0.536   | 382.286  |
|                         | SBERT-specter      |             | 0.929      | 0.630    | 0.500   | 433.143  |
|                         | TFIDF              | Naïve Bayes | 0.964      | 0.723    | 0.714   | 263.464  |
| AtypicalAntipsychotics  | SBERT-mpnet        | Similarity  | 0.786      | 0.169    | 0.172   | 319.731  |
|                         | SBERT-specter      |             | 0.793      | 0.078    | 0.138   | 347.586  |
|                         | TFIDF              | Naïve Bayes | 0.869      | 0.168    | 0.255   | 275.124  |
| Bannach-Brown_2019      | SBERT-mpnet        | Similarity  | 0.968      | 0.550    | 0.423   | 268.194  |
|                         | SBERT-specter      |             | 0.964      | 0.564    | 0.401   | 279.616  |
|                         | TFIDF              | Naïve Bayes | 0.932      | 0.404    | 0.423   | 293.022  |
| BetaBlockers            | SBERT-mpnet        | Similarity  | 0.829      | 0.059    | 0.512   | 487.976  |
|                         | SBERT-specter      |             | 0.805      | 0.090    | 0.537   | 439.415  |
|                         | TFIDF              | Naïve Bayes | 0.927      | 0.669    | 0.610   | 241.878  |
| Bos_2018                | SBERT-mpnet        | Similarity  | 1.000      | 0.772    | 0.800   | 308.300  |
|                         | SBERT-specter      |             | 1.000      | 0.758    | 0.800   | 194.800  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.834    | 0.900   | 76.000   |
| CalciumChannelBlockers  | SBERT-mpnet        | Similarity  | 0.717      | 0.165    | 0.071   | 407.798  |
|                         | SBERT-specter      |             | 0.697      | 0.160    | 0.131   | 419.444  |
|                         | TFIDF              | Naïve Bayes | 0.899      | 0.229    | 0.172   | 268.364  |
| Estrogens               | SBERT-mpnet        | Similarity  | 0.937      | 0.404    | 0.101   | 101.304  |
|                         | SBERT-specter      |             | 0.835      | 0.331    | 0.114   | 107.114  |
|                         | TFIDF              | Naïve Bayes | 0.911      | 0.249    | 0.177   | 89.418   |
| Hall_2012               | SBERT-mpnet        | Similarity  | 1.000      | 0.913    | 0.903   | 99.204   |
|                         | SBERT-specter      |             | 1.000      | 0.892    | 0.903   | 154.184  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.911    | 0.893   | 134.272  |
| Kitchenham_2010         | SBERT-mpnet        | Similarity  | 1.000      | 0.569    | 0.364   | 258.295  |
|                         | SBERT-specter      |             | 1.000      | 0.602    | 0.591   | 160.000  |
|                         | TFIDF              | Naïve Bayes | 0.977      | 0.652    | 0.568   | 200.136  |
| Kwok_2020               | SBERT-mpnet        | Similarity  | 0.975      | 0.507    | 0.529   | 298.529  |
|                         | SBERT-specter      |             | 0.908      | 0.337    | 0.513   | 357.378  |
|                         | TFIDF              | Naïve Bayes | 0.992      | 0.675    | 0.588   | 227.092  |
| Nagtegaal_2019          | SBERT-mpnet        | Similarity  | 0.970      | 0.546    | 0.530   | 236.640  |
|                         | SBERT-specter      |             | 0.950      | 0.447    | 0.500   | 272.790  |
|                         | TFIDF              | Naïve Bayes | 0.990      | 0.529    | 0.540   | 237.190  |
| NSAIDS                  | SBERT-mpnet        | Similarity  | 0.950      | 0.627    | 0.425   | 55.000   |
|                         | SBERT-specter      |             | 0.975      | 0.621    | 0.400   | 51.975   |
|                         | TFIDF              | Naïve Bayes | 0.975      | 0.698    | 0.500   | 46.775   |
| Opiods                  | SBERT-mpnet        | Similarity  | 1.000      | 0.496    | 0.000   | 419.000  |
|                         | SBERT-specter      |             | 0.929      | 0.480    | 0.000   | 396.929  |
|                         | TFIDF              | Naïve Bayes | 0.929      | 0.467    | -0.071  | 717.857  |
| OralHypoglycemics       | SBERT-mpnet        | Similarity  | 0.711      | 0.146    | 0.089   | 182.904  |
|                         | SBERT-specter      |             | 0.711      | 0.104    | 0.119   | 178.896  |
|                         | TFIDF              | Naïve Bayes | 0.770      | 0.190    | 0.141   | 155.444  |
| ProtonPumpInhibitors    | SBERT-mpnet        | Similarity  | 0.880      | 0.290    | 0.240   | 328.520  |
|                         | SBERT-specter      |             | 0.900      | 0.382    | -0.060  | 492.360  |
|                         | TFIDF              | Naïve Bayes | 0.940      | 0.481    | 0.560   | 191.960  |
| Radjenovic_2013         | SBERT-mpnet        | Similarity  | 1.000      | 0.837    | 0.830   | 192.489  |
|                         | SBERT-specter      |             | 1.000      | 0.818    | 0.809   | 266.702  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.841    | 0.851   | 189.574  |
| SkeletalMuscleRelaxants | SBERT-mpnet        | Similarity  | 0.375      | -0.169   | -0.125  | 1066.375 |
|                         | SBERT-specter      |             | 0.625      | -0.042   | -0.125  | 856.500  |
|                         | TFIDF              | Naïve Bayes | 0.625      | -0.015   | -0.125  | 782.625  |
| Statins                 | SBERT-mpnet        | Similarity  | 0.857      | 0.352    | 0.464   | 630.238  |
|                         | SBERT-specter      |             | 0.857      | 0.269    | 0.571   | 589.679  |
|                         | TFIDF              | Naïve Bayes | 0.940      | 0.471    | 0.631   | 445.036  |
| Triptans                | SBERT-mpnet        | Similarity  | 1.000      | 0.553    | 0.348   | 98.696   |
|                         | SBERT-specter      |             | 0.957      | 0.436    | 0.565   | 89.565   |
|                         | TFIDF              | Naïve Bayes | 0.957      | 0.520    | 0.478   | 98.565   |
| UrinaryIncontinence     | SBERT-mpnet        | Similarity  | 0.846      | 0.151    | 0.282   | 79.128   |
|                         | SBERT-specter      |             | 0.846      | 0.089    | 0.256   | 79.744   |
|                         | TFIDF              | Naïve Bayes | 0.872      | 0.382    | 0.359   | 60.923   |
| van_de_Schoot_2017      | SBERT-mpnet        | Similarity  | 0.976      | 0.524    | 0.690   | 511.857  |
|                         | SBERT-specter      |             | 0.976      | 0.599    | 0.762   | 383.929  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.888    | 0.905   | 101.571  |
| van_Dis_2020            | SBERT-mpnet        | Similarity  | 0.986      | 0.622    | 0.306   | 1596.208 |
|                         | SBERT-specter      |             | 0.944      | 0.472    | 0.306   | 1972.181 |
|                         | TFIDF              | Naïve Bayes | 0.972      | 0.650    | 0.556   | 1191.264 |
| Wahono_2015             | SBERT-mpnet        | Similarity  | 1.000      | 0.828    | 0.852   | 239.311  |
|                         | SBERT-specter      |             | 1.000      | 0.803    | 0.803   | 304.344  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.863    | 0.869   | 197.803  |
| Wolters_2018            | SBERT-mpnet        | Similarity  | 1.000      | 0.815    | 0.778   | 126.944  |
|                         | SBERT-specter      |             | 1.000      | 0.823    | 0.833   | 141.889  |
|                         | TFIDF              | Naïve Bayes | 1.000      | 0.794    | 0.778   | 162.222  |

**Reply by rohitgarud** [reply]
Average values excluding Skeletal Muscle Relaxants.
| Feature Extraction | Classifier  | recall_0.5 | wss_0.95 | erf_0.1 | atd     |
|--------------------|-------------|------------|----------|---------|---------|
| SBERT-mpnet        | Similarity  | 0.928      | 0.506    | 0.464   | 307.830 |
| SBERT-specter      |             | 0.913      | 0.468    | 0.469   | 327.193 |
| TFIDF              | Naïve Bayes | 0.945      | 0.562    | 0.544   | 239.400 |

**Reply by rohitgarud** [reply]
Although SBERT-specter performs better than default SBERT-mpnet in the case of a few datasets, the overall performance of 'mpnet' is better than the 'specter' transformer model for SBERT feature extraction.

**Reply by jteijema** [reply]
@rohitgarud 

Thank you for sharing these results! I appreciate the time and effort you've put into this simulation study. It's interesting to see the performance comparison between the SBERT-mpnet and SBERT-specter models. I think the SPECTER transformer has potential for specific use cases, especially when dealing with scientific literature. I'll definitely keep it in mind for future simulations I'll be running. Thanks!

**Reply by rohitgarud** [reply]
Thank you @jteijema .
Actually, I was hoping to get higher accuracy with SPECTER as it is trained on Scientific Papers. Maybe training the SPECTER model on the SLR dataset itself might give a boost to WSS. However, generating the citation-informed dataset is a little difficult.

**Reply by Rensvandeschoot** [reply]
@rohitgarud and what about using the ATD instead of the WSS@95 as performance metric?

**Reply by rohitgarud** [reply]
From the various simulations I have performed, my intuition is WSS@95 is a better metric because many times 95% of records are retrieved quickly but the last 5% of records take longer increasing the ATD. Some models struggle to find the last few records but quickly find 95% of the records.  Due t this ATD does not give as good an indication of model performance as WSS@95, because the ultimate goal is to save the work for total screening. This is just my understanding and I might be missing something. 
Generally, we will look at both values. ATD is still a good metric as we also want to retrieve all the records. If 95% of the records are retrieved quickly and the last 5% take longer to retrieve, they might be missed due to the stopping criteria based on the number of consecutive irrelevant records. Would love to understand your thought process @Rensvandeschoot.

**Reply by Rensvandeschoot** [reply]
I think the Average Time to Discovery (ATD) metric might help compare the performance of different models across different datasets, as it measures performance throughout the entire screening process and does not rely on arbitrary cut-off values. And indeed some hard-to-find-papers might influence its performance. Therefore, the Time to Discovery (TD) metric can be used to identify hard-to-find papers, after which content experts can then evaluate their labels. There is enough work to do for understanding the exact pro's and con's of each metric... it would be an entire PhD project!!!

**Reply by rohitgarud** [reply]
Thank you for response @Rensvandeschoot. I agree, each metric has its own merits. I will keep this in mind while presenting future simulation studies. What I have observed is that the hard-to-find records are sometimes different for different feature/classifier combinations. It is interesting and indeed a topic of research.

**Reply by rohitgarud** [reply]
I think fine-tuned SBERT models has a potential to achieve state-of-the-art results for all of the benchmark datasets. It is really interesting to see how a simple TFIDF/NB combination performs so well outperforming advanced transformer models in case of many datasets and truncation in SBERT might be one of the reasons.

**Reply by Rensvandeschoot** [reply]
great work! see the new datasets in https://github.com/asreview/synergy-dataset

**Comment by rohitgarud** [comment]
One thing to note is that the maximum sequence length of the mpnet model is 384 and that for specter is 512. So in the case of mpnet, if the number of tokens for the text (title+abstract) is more than 384, they will be truncated and will not be used for obtaining the embeddings/feature vector.

**Reply by Rensvandeschoot** [reply]
yes... so, we might need a model which takes more tokens into account... or make sure the most important text is presented first.

**Reply by rohitgarud** [reply]
Yes, @Rensvandeschoot.  I have tried different models such as Longformer (max_seq_length=4096) and other approaches such as chunking the text, generating embedding for the chunk, and then pooling, however, it did not perform well on a few of the datasets I tried this approach on. I also tried the approach mentioned in the SPECTER paper for generating document embeddings using SBERT, but that also did not perform well for a few datasets I tried with. More comprehensive simulation study and exploration of possible approaches is required.

**Reply by Rensvandeschoot** [reply]
Great exploration!
